### PR TITLE
rtp推流增加PCM音频的处理

### PR DESCRIPTION
**Pull request by rayjay214**

### CODE REVIEW

File: src/Rtp/GB28181Process.cpp [hunk 1]
```diff
@@ -102,6 +102,16 @@ bool GB28181Process::inputRtp(bool, const char *data, size_t data_len) {
                 _rtp_decoder[pt] = Factory::getRtpDecoderByCodecId(track->getCodecId());
                 break;
             }
+            case Rtsp::PT_L16_Mono: {
```
Comment (xia-chu) [inline src/Rtp/GB28181Process.cpp:105]: PT_L16_Mono 好像是44100的采样率，你这里固定为16000 是不是有问题？

Comment (rayjay214) [reply src/Rtp/GB28181Process.cpp:105]: 是的，这里没注意，因为之前我们这边固件端推PCM的时候，pt采用的是自定义的97，采样率也是固定的16000，我当时就这样写了，case也用的是97。后来在pr里面就借用zlm里面已有的一个枚举值L16_Mono放在case分支里了

File: src/Rtp/GB28181Process.cpp
Comment (alexliyu7352) [inline src/Rtp/GB28181Process.cpp:1]: AI Review for src/Rtp/GB28181Process.cpp:

# Code Review: Patch to src/Rtp/GB28181Process.cpp

## Summary

This patch adds support for L16 audio RTP payloads in the `GB28181Process` class. It introduces a new case in the `switch` statement to handle L16 RTP packets, creating a new `RtpReceiverImp` instance and adding a corresponding track to the media sink interface.

## Detailed Feedback

### Code Overview

The patch adds support for L16 audio RTP payloads by introducing a new case in the `switch` statement that handles the `Rtsp::PT_L16_Mono` payload type. This case creates a new `RtpReceiverImp` instance with a sample rate of 16000 Hz and adds a corresponding track to the media sink interface.

### Strengths

- **Clear and concise code:** The patch is well-written and easy to understand.
- **Correct handling of L16 audio:** The code correctly handles the L16 audio payload type and creates the appropriate track.
- **Consistent with existing code:** The patch follows the existing coding style and naming conventions.

### Areas for Improvement

#### 1. Missing Documentation

- **Issue:** The new case for L16 audio lacks documentation.
- **Suggestion:** Add a comment explaining the purpose of the new case and the specific handling of L16 audio.
- **Example:**
  ```cpp
  case Rtsp::PT_L16_Mono: {
      // Handle L16 audio RTP packets.
      // Create a new RtpReceiverImp instance with a sample rate of 16000 Hz.
      // Add a corresponding track to the media sink interface.
      ref = std::make_shared<RtpReceiverImp>(16000, [this](RtpPacket::Ptr rtp) { onRtpSorted(std::move(rtp)); });
      auto track = Factory::getTrackByCodecId(CodecL16, 16000, 1, 16);
      CHECK(track);
      track->setIndex(pt);
      _interface->addTrack(track);
      _rtp_decoder[pt] = Factory::getRtpDecoderByCodecId(track->getCodecId());
      break;
  }
  ```

#### 2. Potential for Redundancy

- **Issue:** The code creates a new `RtpReceiverImp` instance for each payload type, which might lead to redundancy if multiple tracks use the same sample rate.
- **Suggestion:** Consider using a map to store `RtpReceiverImp` instances based on sample rate, avoiding unnecessary object creation.
- **Example:**
  ```cpp
  std::map<int, RtpReceiverImp::Ptr> _rtp_receivers;

  // ...

  case Rtsp::PT_L16_Mono: {
      // Handle L16 audio RTP packets.
      auto sample_rate = 16000;
      auto it = _rtp_receivers.find(sample_rate);
      if (it == _rtp_receivers.end()) {
          // Create a new RtpReceiverImp instance if it doesn't exist.
          ref = std::make_shared<RtpReceiverImp>(sample_rate, [this](RtpPacket::Ptr rtp) { onRtpSorted(std::move(rtp)); });
          _rtp_receivers[sample_rate] = ref;
      } else {
          // Reuse existing RtpReceiverImp instance.
          ref = it->second;
      }
      // ...
      break;
  }
  ```

#### 3. Missing Test Cases

- **Issue:** The patch lacks unit tests to verify the functionality of the new L16 audio handling.
- **Suggestion:** Add unit tests to ensure that the code correctly handles L16 audio RTP packets and creates the appropriate track.

## Conclusion

This patch effectively adds support for L16 audio RTP payloads. However, it lacks documentation and could be improved by reducing redundancy and adding unit tests. Addressing these points will enhance the code's clarity, maintainability, and reliability.

`TRANS_BY_GITHUB_AI_ASSISTANT`

File: src/Rtsp/RtpReceiver.h
Comment (alexliyu7352) [inline src/Rtsp/RtpReceiver.h:1]: AI Review for src/Rtsp/RtpReceiver.h:

# Code Review: Patch to src/Rtsp/RtpReceiver.h

## Summary

This patch modifies the `PacketSortor` class to improve its handling of out-of-order RTP packets, particularly when dealing with potential sequence number resets by the sender.  It introduces a new cache (`_pkt_drop_cache_map`) to temporarily store packets with sequence numbers lower than the expected next sequence number.  The changes aim to reduce packet loss and improve robustness in scenarios with unreliable network conditions.  This is a patch review.

## Detailed Feedback

### Code Overview

The patch enhances the `PacketSortor` class by adding a mechanism to handle situations where the RTP sender might reset its sequence number counter.  It does this by buffering packets with lower sequence numbers than expected and flushing them under certain conditions (buffer size exceeded, time elapsed, or too many dropped packets).

### Strengths

- **Improved Out-of-Order Packet Handling:** The addition of `_pkt_drop_cache_map` directly addresses the problem of sequence number resets, a common issue in RTP streams.
- **More Robustness:** The logic for flushing the cache based on buffer size, time, and the number of dropped packets makes the system more resilient to network fluctuations.
- **Clearer Logic (partially):** The separation of packet handling into `_pkt_sort_cache_map` and `_pkt_drop_cache_map` improves the conceptual clarity of the code.

### Areas for Improvement

#### 1.  Clarity and Readability

- **Issue:** The logic within `forceFlush` and the interaction between `_pkt_sort_cache_map` and `_pkt_drop_cache_map` could be made clearer. The comments are helpful but don't fully explain the complex logic.
- **Suggestion:** Consider restructuring the `forceFlush` function to break down the steps more explicitly.  Add more detailed comments explaining the conditions under which each part of the logic executes.  Consider using more descriptive variable names.  For example, instead of `it`, use `it_lower_bound` or a similar name to indicate its purpose.
- **Example:**

```c++
    void forceFlush(SEQ next_seq) {
        if (_pkt_sort_cache_map.empty()) {
            return;
        }

        // Find the iterator to the first packet with a sequence number >= next_seq
        auto it_lower_bound = _pkt_sort_cache_map.lower_bound(next_seq);

        if (it_lower_bound == _pkt_sort_cache_map.end()) {
            // No packets with seq >= next_seq found; likely a loop or significant packet loss.
            // Process the oldest packet as the next expected packet.
            it_lower_bound = _pkt_sort_cache_map.begin();
        }

        // Process packets up to the found iterator
        popIterator(it_lower_bound);

        // ... (rest of the forceFlush logic with improved comments) ...
    }
```

#### 2.  Error Handling and Logging

- **Issue:** The `WarnL` message in `output` only logs when a packet is dropped.  More comprehensive logging could be beneficial for debugging.
- **Suggestion:** Add logging statements to track the size of the caches, the number of packets dropped, and other relevant metrics.  Consider using different log levels (e.g., `DebugL`, `InfoL`, `WarnL`, `ErrorL`) to categorize the messages appropriately.

#### 3.  Parameter Tuning

- **Issue:** The default value of `_max_buffer_ms` is changed from 1000 to 300.  The rationale for this change is not clear.
- **Suggestion:**  Provide a clear explanation in a comment for why this value was reduced.  Consider adding more comments explaining the trade-offs involved in choosing values for `_max_buffer_size`, `_max_buffer_ms`, and `_max_distance`.  These parameters significantly impact the behavior of the `PacketSortor`.

#### 4.  Testing

- **Issue:**  The patch lacks accompanying unit tests to verify the correctness of the added logic.
- **Suggestion:**  Add unit tests to cover various scenarios, including normal packet arrival, out-of-order packets, sequence number resets, and edge cases (e.g., very large sequence number gaps).


## Conclusion

The patch introduces a valuable improvement to the RTP packet sorting mechanism.  However, improvements in clarity, logging, parameter explanation, and testing are needed to ensure its robustness and maintainability.  Addressing these points will significantly enhance the quality of the code.

`TRANS_BY_GITHUB_AI_ASSISTANT`
